tests/framework/builder.py

```diff
@@ -40,11 +54,88 @@ def build(self, kernel: Artifact, disks: [DiskArtifact], config: Artifact):
+        jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
+        assert len(snapshot.disks) > 0, "Snapshot requiures at least one disk."
+        _jailed_disk = vm.create_jailed_resource(snapshot.disks[0])
+        vm.ssh_config['ssh_key_path'] = snapshot.ssh_key
+
+        vm.create_tap_and_ssh_config(host_ip=host_ip,
+                                     guest_ip=guest_ip,
+                                     netmask_len=netmask_len,
+                                     tapname="tap0")
+
+        response = vm.snapshot_load.put(mem_file_path=jailed_mem,
+                                        snapshot_path=jailed_vmstate,
+                                        diff=enable_diff_snapshots)
+
+        assert vm.api_session.is_status_no_content(response.status_code)
+
+        if resume:
+            # Resume microvm
+            response = vm.vm.patch(state='Resumed')
+            assert vm.api_session.is_status_no_content(response.status_code)
+
+        # Return a resumed microvm.
+        return vm
+
+
+class SnapshotBuilder:  # pylint: disable=too-few-public-methods
+    """Create a snapshot from a running microvm."""
+
+    def __init__(self, microvm):
+        """Initialize the snapshot builder."""
+        self._microvm = microvm
+
+    def create(self,
+               disks: [DiskArtifact],
+               ssh_key: Artifact,
+               snapshot_type: SnapshotType = SnapshotType.FULL):
+        """Create a Snapshot object from a microvm and artifacts."""
+        # Disable API timeout as the APIs for snapshot related procedures
+        # take longer.
+        self._microvm.api_session.untime()
+        chroot_path = self._microvm.jailer.chroot_path()
+        snapshot_dir = os.path.join(chroot_path, "snapshot")
+        Path(snapshot_dir).mkdir(parents=True, exist_ok=True)
+        cmd = 'chown {}:{} {}'.format(self._microvm.jailer.uid,
+                                      self._microvm.jailer.gid,
+                                      snapshot_dir)
+        utils.run_cmd(cmd)
+        self._microvm.pause_to_snapshot(
+            mem_file_path="/snapshot/vm.mem",
+            snapshot_path="/snapshot/vm.vmstate",
+            diff=snapshot_type == SnapshotType.DIFF)
+
+        # Create a copy of the ssh_key artifact.
+        ssh_key_copy = ssh_key.copy()
+        mem_path = os.path.join(snapshot_dir, "vm.mem")
+        vmstate_path = os.path.join(snapshot_dir, "vm.vmstate")
+        return Snapshot(mem=mem_path,
+                        vmstate=vmstate_path,
+                        # TODO: To support more disks we need to figure out a
+                        # simple and flexible way to store snapshot artifacts
+                        # in S3. This should be done in a PR where we add tests
+                        # that resume from S3 snapshot artifacts.
+                        disks=[disks[0].local_path()],
+                        ssh_key=ssh_key_copy.local_path())
```

Comment:
It doesn't need to be writeable but also does not hurt.

Reply:
If it is not needed, we should not use it, even if it does not hurt. Keeping it just complicates the code.

Reply:
Can you explain how it complicates the code ? Actually we need to add more code select between read only or writeable copy.

Reply:
To me seems redundant that we `copy` the `ssh_key`, then use the copy local path for the snapshot, instead of using directly the `ssh_key` local_path.